extensions/wikia/MercuryApi/MercuryApiController.class.php

```diff
@@ -405,41 +282,57 @@ public function getPage() {
 			$article = Article::newFromID( $title->getArticleId() );
 
 			if ( $article instanceof Article && $title->isRedirect() ) {
-				list( $title, $article, $data ) =
-					$this->handleRedirect( $title, $article, $data );
+				list( $title, $article, $data ) = $this->handleRedirect( $title, $article, $data );
 			}
 
 			$isMainPage = $title->isMainPage();
 			$data['isMainPage'] = $isMainPage;
-
-			$otherLanguages = $this->getOtherLanguages( $title );
-
-			if ( !empty( $otherLanguages ) ) {
-				$data['otherLanguages'] = $otherLanguages;
-			}
+			$data['ns'] = $title->getNamespace();
 
 			$titleBuilder = new WikiaHtmlTitle();
+			if ( MercuryApiMainPageHandler::shouldGetMainPageData( $isMainPage ) ) {
+				$data['mainPageData'] = MercuryApiMainPageHandler::getMainPageData( $this->mercuryApi );
 
-			if ( $this->shouldGetMainPageData( $isMainPage ) ) {
-				$data['mainPageData'] = $this->getMainPageData();
 				if ( $article instanceof Article ) {
-					$data['details'] = $this->getArticleDetails( $article );
+					$data['details'] = MercuryApiArticleHandler::getArticleDetails( $article );
 				} else {
-					$data['details'] = $this->getMainPageMockedDetails( $title );
+					$data['details'] = MercuryApiMainPageHandler::getMainPageMockedDetails( $title );
 				}
-			} else if ( $title->isContentPage() && $title->isKnown() ) {
-				$data = array_merge( $data, $this->getArticleData( $article ) );
-
-				if ( !$isMainPage ) {
-					$titleBuilder->setParts( [ $data['article']['displayTitle'] ] );
+			} else {
+				switch ( $data['ns'] ) {
+					// Handling namespaces other than content ones
+					case NS_CATEGORY:
+						$data['nsSpecificContent'] = MercuryApiCategoryHandler::getCategoryContent( $title );
+
+						if ( MercuryApiCategoryHandler::hasArticle( $title ) ) {
+							$data['article'] = MercuryApiArticleHandler::getArticleJson( $this->request, $article );
+							$data['details'] = MercuryApiArticleHandler::getArticleDetails( $article );
+						} elseif ( !empty( $data['nsSpecificContent']['members']['sections'] ) ) {
+							$data['details'] = MercuryApiCategoryHandler::getCategoryMockedDetails( $title );
+						} else {
+							throw new NotFoundApiException( 'Article is empty and category has no members' );
+						}
+						break;
+					// Handling content namespaces
+					default:
+						if ( $title->isContentPage() && $title->isKnown() && $article instanceof Article ) {
+							$data = array_merge(
+								$data,
+								MercuryApiArticleHandler::getArticleData( $this->request, $this->mercuryApi, $article )
+							);
+
+							if ( !$isMainPage ) {
+								$titleBuilder->setParts( [ $data['article']['displayTitle'] ] );
+							}
+						} else {
```

Comment:
I'm a little bit worried here.
Maybe this exception should be throw only if article isn't an instance of Article.
Right now it would throw in case of no-content pages and not known titles and articles with id 0.